broker/protocol/dispatcher.go

```diff
@@ -117,12 +117,19 @@ func (d *dispatcher) UpdateSubConnState(sc balancer.SubConn, state balancer.SubC
 		panic("unexpected SubConn")
 	}
 
-	if state.ConnectivityState == connectivity.Connecting && d.connState[sc] == connectivity.TransientFailure {
+	if (state.ConnectivityState == connectivity.Connecting || state.ConnectivityState == connectivity.Idle) &&
+		d.connState[sc] == connectivity.TransientFailure {
 		// gRPC will quickly transition failed connections back into a Connecting
 		// state. In many cases, such as a remote-initiated close from a
 		// shutting-down server, the SubConn may never return. Until we see a
 		// successful re-connect, continue to consider the SubConn as broken
 		// (and trigger invalidations of cached Routes which use it).
+
+		if state.ConnectivityState == connectivity.Idle {
+			sc.Connect()
+		}
+		d.mu.Unlock()
+		return
```

Comment:
Why is it important to return here, where we don't return with the Connect() call below?
As far as I can tell, the purpose is to _not_ call UpdateState below. Is that true, and why would it a problem to do so?

(I'm trying to figure out if this code block could be eliminated, and instead have the check below be the only place where Connect() is called).

Reply:
Also, the core problem (as I understand it) is that Idle sub-connections won't actually re-connect unless `sc.Connect()` is explicitly called, right?

Was this a change in gRPC behavior? I'm confused how this isn't a problem we've experienced ourselves. For that matter, help me understand: how does a previously-active SubConn come to be Idle again (my very old recollection is they bounce between failure and `Connecting`)?

In any case, since the check below will cause Connect to be called, is there a particular reason that we don't want to update `connState[sc]` if it's `state.Idle` now?

Put another way, would the root issue would be resolve with just the addition below of:
```
	if state.ConnectivityState == connectivity.Idle {
		sc.Connect()
	}
```

Almost certainly I'm missing something 🤷 .

Reply:
No problem. I am going to re-look into the changes that occurred in gRPC over time, as it has been updated in our master a couple of times, to see if it plays any part. Agree that it is strange it has not popped up in your deployment. 

A lot of my investigation was the evolution (via blame) in https://github.com/grpc/grpc-go/blob/master/balancer/base/balancer.go#L180 as we appeared to be lined up pretty closely with the overall pattern of UpdateSubConnState() but they have been tweaking it over time. 

I will get back with hopefully good answers to your questions and your simplification may be valid.

Reply:
Still looking but i think this gRPC PR https://github.com/grpc/grpc-go/commit/03268c8ed29e801944a2265a82f240f7c0e1b1c3#diff-f9772420643575b997f617c6d4c1934aaa26f057042c7fa71a521ef5bb2af253 is the one where gRPC have introduced the transition to the Idle state and adjusted their own example loadbalancer to work with that. 

In their example balancer/base/balancer.go they were returning in the top check for Idle state and bypassing the call to UpdateState() so i followed that. I can dig deeper on that to see why that may or may not make a difference.

On thinking on why we were seeing this issue and you were not it may be related to how we run our cluster (not sure). Our brokers do restart independent of the consumers (autoscaling and some spot instances). There are also times that our gitops methodology might restart the gazette brokers and not all the consumers. We do find that in a controlled rollout where the brokers first restart and then we restart the consumers there has always been full connectivity with optimal zonal wiring. This PR just helps in the case where the brokers restart and the consumers do not.

Will look around a bit more.

Reply:
It seems like the gRPC change above was for the 1.41.0 release which the gazette repo does not yet use. So this PR may be premature.

Reply:
Will talk to Michael when he comes back next week to see if we should close this PR to master.

Reply:
@jgraettinger i guess these(or similar) changes will be required when  gRPC is upgraded to 1.41+. For now i changed the title to reflect this. I will leave open for now but will close if you think that is best.